### ISSUE TITLE
Mapping strings from Azure website into app code

### ISSUE THREAD

**Issue by jlinstrum** [dotnet/AspNetCore.Docs#17913]
I managed to get the blazorwasm template working, and it only took me about a day to do it, which I don't consider to be too bad. Aside from the large number of steps (unavoidable) the part that was most difficult was taking the information written down while creating the AAD B2C resources on the website and mapping that info into the terms for the "dotnet new" command (or trying to map it to the manual configuration, which I gave up on).

In particular, it was challenging to figure out what "DOMAIN" meant as there seemed to be more than one candidate. It's probably obvious to someone with more understanding of the subject - me, not so much. I made some mistakes and had to do some trial and error to get it working.

Thanks for the document, it was very helpful. If possible, a table more clearly mapping the information from the AAD B2C website to the {TERMS} in the "dotnet new" command and the configuration documentation would be amazing. :)

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7d682d4c-86b5-c47e-89a3-3684b2235320
* Version Independent ID: a80e115c-466a-288d-a918-9ef7514f5add
* Content: [Secure an ASP.NET Core Blazor WebAssembly hosted app with Azure Active Directory B2C](https://docs.microsoft.com/en-us/aspnet/core/security/blazor/webassembly/hosted-with-azure-active-directory-b2c?view=aspnetcore-3.1#feedback)
* Content Source: [aspnetcore/security/blazor/webassembly/hosted-with-azure-active-directory-b2c.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/security/blazor/webassembly/hosted-with-azure-active-directory-b2c.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
Yes, I'll take that under consideration. I split the entries and used bullets. For example, it says for the tenant ...

> ... record the following information:
> * AAD B2C instance (for example, `https://contoso.b2clogin.com/`, which includes the trailing slash)
> * AAD B2C Tenant domain (for example, `contoso.onmicrosoft.com`)

Then, for the server API, it says ...

> Record the following information:
>
> * Server API app Application ID (Client ID) (for example, 11111111-1111-1111-1111-111111111111)
> * App ID URI (for example, https://contoso.onmicrosoft.com/11111111-1111-1111-1111-111111111111, api://11111111-1111-1111-1111-111111111111, or the custom value that you provided)
> * Directory ID (Tenant ID) (for example, 222222222-2222-2222-2222-222222222222)
> * Server API app App ID URI (for example, https://contoso.onmicrosoft.com/11111111-1111-1111-1111-111111111111, the Azure portal might default the value to the Client ID)
> * Default scope (for example, API.Access)

Finally for the client app, it says ...

> Record the following information:
>
> * Record the Client app Application ID (Client ID) (for example, 33333333-3333-3333-3333-333333333333).
> * Record the sign-up and sign-in user flow name created for the app (for example, B2C_1_signupsignin).

I'll take a look at a few things:

* Keying the values better to the command that follows.
* Possibly using a clear table of values (or three of them maybe).

**Comment by Xaeco** [comment]
I had trouble with the "AAD B2C instance" bit. I looked through all the B2C screens but there is no reference to it. I assumed it was my {b2c name}.b2clogin.com which worked.

**Comment by manasimahajann** [comment]
I am wondering when I am creating this app - 
I get redirected to "The site can't be reached" with the below in the url
https://localhost:5001/authentication/login-callback#error=redirect_uri_mismatch&error_description=AADB2C90006%3a+The+redirect+URI+%27https%3a%2f%2flocalhost%3a44344%2fauthentication%2flogin-callback%27+provided+in+the+request+is+not+registered+for+the+client+id+

**Comment by guardrex** [comment]
@manasimahajann Looks like you didn't register the callback in the Azure portal. It's covered in the docs. If you need support, try the support channels ...

* [Stack Overflow](https://stackoverflow.com/questions/tagged/blazor)
* [ASP.NET Core Slack](http://tattoocoder.com/aspnet-slack-sign-up/)
* [ASP.NET Gitter](https://gitter.im/aspnet/Blazor)

**Comment by manasimahajann** [comment]
thanks for your response, we changed our redirect url to the one in the error 44344 and it worked.
Is there a way we could be able to add more than on scope, i tried list or commas or even adding the same line over and over with new scope- gives an error
 options.ProviderOptions.DefaultAccessTokenScopes.Add()

**Comment by guardrex** [comment]
I've run into my own pain with the process for B2C. I needed an additional scope for Graph API so that I could make Azure AD role and group claims in the app, but I was never able to get a token for the Graph API call ...

https://github.com/dotnet/AspNetCore.Docs/issues/17683#issuecomment-627498149

I don't know if it's supported at this time.

**Comment by Xaeco** [comment]
Have a look at [https://docs.microsoft.com/en-us/aspnet/core/security/blazor/webassembly/additional-scenarios?view=aspnetcore-3.1#request-additional-access-tokens](url)

**Comment by manasimahajann** [comment]
Xaeco,
Thank you. Tried to get to that link.. We get, 404 - page not found

**Comment by guardrex** [comment]
That process doesn't seem to work. See my notes at https://github.com/dotnet/AspNetCore.Docs/issues/17683#issuecomment-627498149. I'm either missing something (I have a 🐞 in my bits), or the scenario isn't supported at this time.

If you get it working, please let me know what you did. I had to drop that work temporarily to turn to higher priority issues, so I didn't have a chance to investigate further.

**Comment by manasimahajann** [comment]
guardex,
I just checked that when I was trying to get to fetch data -- I am not getting any info -- it says "Loading"

When i inspect the page below is the error, let me know if you have come across this 

blazor.webassembly.js:1 WASM: ﻿Unhandled exception rendering component:
p.printErr @ blazor.webassembly.js:1
blazor.webassembly.js:1 WASM: System.NullReferenceException: Object reference not set to an instance of an object.
p.printErr @ blazor.webassembly.js:1
blazor.webassembly.js:1 WASM:   at newAuther.Client.Pages.FetchData.OnInitializedAsync () [0x000e6] in C:\dev\Blazor\newAuther\Client\Pages\FetchData.razor:54 
p.printErr @ blazor.webassembly.js:1
blazor.webassembly.js:1 WASM:   at Microsoft.AspNetCore.Components.ComponentBase.RunInitAndSetParametersAsync () <0x2a691c0 + 0x0014c> in <filename unknown>:0 
p.printErr @ blazor.webassembly.js:1
blazor.webassembly.js:1 WASM:   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask (System.Threading.Tasks.Task taskToHandle) <0x2cd5668 + 0x000c2> in <filename unknown>:0

**Comment by guardrex** [comment]
That's a problem with the default scope (probably). If you tried adding multiple scopes with different API endpoints, I think that's normal for it to fail like that. Additional scopes (for AAD anyway) to a different API are supposed to be added as **_additional scopes_** (`AdditionalScopesToConsent`) AFAIK.

My app does make regular API requests for the weather data. My `FetchData` component and server API work fine. My problem is when I try to set up the app to make Graph API calls **_in addition to the server API for weather data_**. My additional scopes case doesn't work in B2C, but it does work in AAD. It might just not be supported for B2C at this time. I'll need feedback from engineering (or the community) to proceed ... or figure it out with more research when I can get back to it.

**Comment by manasimahajann** [comment]
@guardex, with default scope - only one -- i am getting that error , I wonder if token is not reaching the server

**Comment by guardrex** [comment]
Thanks @manasimahajann ... We'll get some help with this soon. It should be seen by Friday. If not, I'll ping internally for engineering to have a look at this issue and my notes on the other issue.

**Comment by manasimahajann** [comment]
Sure , thank you.

**Comment by Xaeco** [comment]
> I am wondering when I am creating this app -
> I get redirected to "The site can't be reached" with the below in the url
> https://localhost:5001/authentication/login-callback#error=redirect_uri_mismatch&error_description=AADB2C90006%3a+The+redirect+URI+%27https%3a%2f%2flocalhost%3a44344%2fauthentication%2flogin-callback%27+provided+in+the+request+is+not+registered+for+the+client+id+

I had the same problem so 

> Xaeco,
> Thank you. Tried to get to that link.. We get, 404 - page not found

I’ve updated the link. Hope it helps

**Comment by manasimahajann** [comment]
@guardrex Hey, we were able to get the authentication to work alright.
In the long string (to use it from command line to create a project), we made one change. For --app-id-uri : there we put server's client id.
Also, in our naming convention - we had removed onmicrosoft.com from certain places - which made it difficult to find the issue.
One question about the authentication, in fetchdata.razor we are creating a token - could we be able to do that in program.cs in the client side? so, it need not be created on every screen.I am not able to initiate **IAccessTokenProvider** in the program.cs of client project - any ideas?
